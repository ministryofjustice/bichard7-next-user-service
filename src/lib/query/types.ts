export type DBResult<T> = T
export type TransformResult<T> = T
export type Transform<I, O> = (dbResult: DBResult<I>) => TransformResult<O>
export type QueryError = Error
export type OnError = (e: QueryError) => void

export enum QueryType {
  Any = "Any"
}
