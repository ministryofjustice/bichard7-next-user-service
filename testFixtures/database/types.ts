export enum Tables {
  Users = 'users'
}