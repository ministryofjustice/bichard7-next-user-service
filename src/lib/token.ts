import jwt from "jsonwebtoken"
import config from "lib/config"
import { UserGroup } from "lib/User"

const signOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

export type Token = string

export interface BichardTokenPayload {
  username: string
  exclusionList: string[]
  inclusionList: string[]
  forenames: string
  surname: string
  emailAddress: string
  groups: UserGroup[]
}

export interface EmailTokenPayload {
  emailAddress: string
  verificationCode: string
}

export function generateBichardToken(payload: BichardTokenPayload): Token {
  const options: jwt.SignOptions = {
    expiresIn: config.tokenExpiresIn,
    ...signOptions
  }

  return jwt.sign(payload, config.tokenSecret, options)
}

export function generateEmailToken(payload: EmailTokenPayload): Token {
  const options: jwt.SignOptions = {
    expiresIn: "3 hours",
    ...signOptions
  }

  return jwt.sign(payload, config.tokenSecret, options)
}
