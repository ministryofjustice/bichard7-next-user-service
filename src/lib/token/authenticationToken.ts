import jwt from "jsonwebtoken"
import { Result } from "types/Result"
import UserFullDetails from "types/UserFullDetails"
import Database from "types/Database"
import PromiseResult from "types/PromiseResult"
import config from "../config"
import UserGroup from "../../types/UserGroup"

const signOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

export type AuthenticationToken = string

export interface AuthenticationTokenPayload {
  username: string
  exclusionList: string[]
  inclusionList: string[]
  emailAddress: string
  groups: UserGroup[]
  id: string
}

export function generateAuthenticationToken(user: Partial<UserFullDetails>, uniqueId: string): AuthenticationToken {
  const options: jwt.SignOptions = {
    expiresIn: config.tokenExpiresIn,
    ...signOptions
  }

  const payload: AuthenticationTokenPayload = {
    username: user.username as string,
    exclusionList: user.exclusionList as string[],
    inclusionList: user.inclusionList as string[],
    emailAddress: user.emailAddress as string,
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    groups: user.groups as UserGroup[],
    id: uniqueId
  }

  return jwt.sign(payload, config.tokenSecret, options)
}

export function decodeAuthenticationToken(token: string): Result<AuthenticationTokenPayload> {
  try {
    return jwt.verify(token, config.tokenSecret, signOptions) as AuthenticationTokenPayload
  } catch (error) {
    return error as Error
  }
}

export async function isTokenIdValid(connection: Database, uniqueId: string): PromiseResult<boolean> {
  /* eslint-disable no-useless-escape */
  const query = `
    SELECT COUNT(1)
    FROM br7own.jwt_ids
    WHERE id = $\{uniqueId\};
  `
  /* eslint-disable no-useless-escape */

  try {
    const count = await connection.one(query, { uniqueId })
    return !!count
  } catch (error) {
    return error as Error
  }
}

export async function removeTokenId(connection: Database, uniqueId: string): PromiseResult<void> {
  /* eslint-disable no-useless-escape */
  const removeTokenIdQuery = `
    DELETE FROM br7own.jwt_ids
    WHERE id = $\{uniqueId\};
  `
  /* eslint-disable no-useless-escape */

  try {
    await connection.none(removeTokenIdQuery, { uniqueId })
    return undefined
  } catch (error) {
    return error as Error
  }
}

export async function storeTokenId(connection: Database, userId: number, uniqueId: string): PromiseResult<void> {
  /* eslint-disable no-useless-escape */
  const storeTokenIdQuery = `
    INSERT INTO br7own.jwt_ids
    (
      id,
      generated_at,
      user_id
    )
    VALUES
    (
      $\{id\},
      NOW(),
      $\{user_id\}
    );
  `
  /* eslint-disable no-useless-escape */

  try {
    await connection.none(storeTokenIdQuery, { id: uniqueId, user_id: userId })
    return undefined
  } catch (error) {
    return error as Error
  }
}
