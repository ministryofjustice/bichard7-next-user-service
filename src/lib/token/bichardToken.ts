import jwt from "jsonwebtoken"
import User from "types/User"
import config from "../config"
import UserGroup from "../../types/UserGroup"

const signOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

export type BichardToken = string

export interface BichardTokenPayload {
  username: string
  exclusionList: string[]
  inclusionList: string[]
  emailAddress: string
  groups: UserGroup[]
}

export function generateBichardToken(user: User): BichardToken {
  const options: jwt.SignOptions = {
    expiresIn: config.tokenExpiresIn,
    ...signOptions
  }

  const payload: BichardTokenPayload = {
    username: user.username,
    exclusionList: user.exclusionList,
    inclusionList: user.inclusionList,
    emailAddress: user.emailAddress,
    groups: user.groups
  }

  return jwt.sign(payload, config.tokenSecret, options)
}
