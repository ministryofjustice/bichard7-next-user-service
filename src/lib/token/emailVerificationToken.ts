import jwt from "jsonwebtoken"
import config from "lib/config"

const signOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

const verifyOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

export type EmailVerificationToken = string

export interface EmailVerificationTokenPayload {
  emailAddress: string
  verificationCode: string
}

export function decodeEmailVerificationToken(token: EmailVerificationToken): EmailVerificationTokenPayload {
  try {
    return jwt.verify(token, config.tokenSecret, verifyOptions) as EmailVerificationTokenPayload
  } catch (error) {
    return error
  }
}

export function generateEmailVerificationToken(payload: EmailVerificationTokenPayload): EmailVerificationToken {
  const options: jwt.SignOptions = {
    expiresIn: `${config.emailVerificationExpiresIn} minutes`,
    ...signOptions
  }

  try {
    return jwt.sign(payload, config.tokenSecret, options)
  } catch (error) {
    return error
  }
}
