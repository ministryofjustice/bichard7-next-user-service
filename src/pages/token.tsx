import ErrorSummary from "components/ErrorSummary"
import GridRow from "components/GridRow"
import Layout from "components/Layout"
import Head from "next/head"
import { GetServerSideProps } from "next"
import Auth from "@aws-amplify/auth"

interface Props {
  failure?: boolean
  token?: string
  tokenContents?: string
}

export const getServerSideProps: GetServerSideProps = async () => {
  const props: Props = {}

  try {
    const user = await Auth.currentAuthenticatedUser()
    const idToken = user.getSignInUserSession()?.getIdToken()
    props.token = idToken?.getJwtToken()
    props.tokenContents = JSON.stringify(idToken?.decodePayload(), null, 4)
  } catch (error) {
    console.log(error)
    props.failure = true
  }

  return { props }
}

const Token = ({ failure, token, tokenContents }: Props) => (
  <>
    <Head>
      <title>{"Sign in to Bichard 7"}</title>
    </Head>
    <Layout>
      {failure && (
        <GridRow>
          <ErrorSummary title="Failed to get token">{"There was a problem getting the token"}</ErrorSummary>
        </GridRow>
      )}

      {token && (
        <GridRow>
          <code style={{ wordWrap: "break-word" }}>{token}</code>
        </GridRow>
      )}

      {tokenContents && (
        <GridRow>
          <pre>{tokenContents}</pre>
        </GridRow>
      )}
    </Layout>
  </>
)

export default Token
