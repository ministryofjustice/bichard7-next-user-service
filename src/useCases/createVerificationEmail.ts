import { EmailVerificationTokenPayload, generateEmailVerificationToken } from "lib/token/emailVerificationToken"
import EmailResult from "types/EmailResult"
import { isError, Result } from "types/Result"

export default (emailAddress: string, verificationCode: string, redirectUrl?: string): Result<EmailResult> => {
  const payload: EmailVerificationTokenPayload = {
    emailAddress,
    verificationCode
  }
  const token = generateEmailVerificationToken(payload)
  if (isError(token)) {
    return token
  }
  const url = new URL("/login/verify", "http://localhost:3000")
  url.searchParams.append("token", token)

  if (redirectUrl) {
    url.searchParams.append("redirect", redirectUrl)
  }

  const subject = "Sign in to Bichard"
  const body = `
    TO: ${emailAddress}

    Click here to log in to Bichard:
    ${url.href}
  `

  return { subject, body }
}
