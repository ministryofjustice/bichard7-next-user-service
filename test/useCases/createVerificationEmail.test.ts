import { isError } from "types/Result"
import EmailContent from "types/EmailContent"
import { generateEmailVerificationToken } from "lib/token/emailVerificationToken"
import createVerificationEmail from "useCases/createVerificationEmail"

jest.mock("lib/token/emailVerificationToken")

it("should generate the email subject and body", () => {
  const mockedGenerateEmailVerificationToken = generateEmailVerificationToken as jest.MockedFunction<
    typeof generateEmailVerificationToken
  >
  mockedGenerateEmailVerificationToken.mockReturnValue("DUMMY_TOKEN")

  const result = createVerificationEmail("bichard01@example.com", "123456")

  expect(isError(result)).toBe(false)

  const email = result as EmailContent
  expect(email.subject).toMatchSnapshot()
  expect(email.text).toMatchSnapshot()
  expect(email.html).toMatchSnapshot()
})
