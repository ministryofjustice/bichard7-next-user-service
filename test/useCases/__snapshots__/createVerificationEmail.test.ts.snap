// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate the email subject and body 1`] = `"Sign in to Bichard"`;

exports[`should generate the email subject and body 2`] = `
"Sign in to Bichard

In order to sign in to Bichard, you need to confirm your email address. Please confirm your email address by clicking the link below.

http://localhost:3000/login/verify?token=DUMMY_TOKEN

If you didn't request this email, you can safely ignore it.
"
`;

exports[`should generate the email subject and body 3`] = `"<html lang=\\"en\\"><head><meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\"/><title>Sign in to Bichard</title></head><body><table style=\\"background-color:#f3f2f1;margin:0;padding:30px;width:100%\\"><tr></tr><tr><td style=\\"background-color:#ffffff;border:1px solid;border-color:#b1b4b6;clear:both;display:block !important;margin:0 auto;max-width:600px !important;padding:30px;vertical-align:top\\" valign=\\"top\\" width=\\"600\\"><h1 style=\\"color:#0b0c0c;font-family:Arial, sans-serif;font-size:48px;line-height:50px;margin-bottom:50px\\">Sign in to Bichard</h1><p style=\\"color:#0b0c0c;font-family:Arial, sans-serif;font-size:19px;line-height:25px;margin-bottom:20px\\">In order to sign in to Bichard, you need to confirm your email address. Please confirm your email address by clicking the button below.</p><p style=\\"color:#0b0c0c;font-family:Arial, sans-serif;font-size:19px;line-height:25px;margin-bottom:20px\\">If you didn&#x27;t request this email, you can safely ignore it.</p><a href=\\"http://localhost:3000/login/verify?token=DUMMY_TOKEN\\" style=\\"background-color:#00703c;color:#ffffff;cursor:pointer;display:inline-block;font-family:Arial, sans-serif;font-size:19px;line-height:19px;padding:8px 10px 7px 10px;text-align:center;text-decoration:none\\">Sign in</a><p style=\\"color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px;margin-top:50px\\">If youâ€™re having trouble clicking the button, copy and paste the URL below into your web browser:</p><p style=\\"color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px;margin:0\\"><a href=\\"http://localhost:3000/login/verify?token=DUMMY_TOKEN\\" style=\\"color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px\\">http://localhost:3000/login/verify?token=DUMMY_TOKEN</a></p></td></tr><tr></tr></table></body></html>"`;
